# Exploring eBay Car Sales Data Project

In this project, I will analyze a data set of used car listings that were scraped from the classified section of German eBay. The data set was obtained from Kaggle and a random sample of 50,000 data points was taken. The data set and documentation, including explanation of all the columns, is [here](https://www.kaggle.com/orgesleka/used-cars-database/data).

## Importing Data and Initial Exploration

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

Above, I imported the NumPy and pandas libraries and read the csv file into a pandas dataframe. Since the default encoding of UTF-8 resulted in an error, I changed the encoding to Latin-1.

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

I can see that there are 20 columns in the data set. Some columns have missing values: vehicleType, gearbox, model, fuelType, and notRepairedDamage. Five of the columns are of the integer data type, and the other 15 columns are of the object (string) data type.

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

I can also see from the first 5 rows that the data is in German, which the Kaggle documentation stated. Angebot, for example, translates to "offer".

## Column Names

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

After looking at the columns, I can see they are in camelcase. I will convert them all to snakecase by specifying new column names in a list and assigning that list to the columns attribute of the data set. I will also modify the column names where possible to make them more clear.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fue_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fue_type  \
0    manuell       158  andere  150,000km                   3      lpg   
1  automatik       286     7er  150,000km                   6   benzin   
2    manuell       102    golf   70,000km                   7   benzin   
3  automatik        71  fortwo   70,000km                   6   benzin   
4    manuell         0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Data Cleaning

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fue_type       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Based on the descriptive statistics above, I can see that there are some columns that contain pretty much all the same values and are therefore not very useful for analysis. These columns are: seller (all except one data point are private party), offer type (all except one data point are Angebot, which translates to "offer"), and number of pictures (all are 0). Price and odometer are stored as text so should be cleaned and changed to a numeric data type. Registration year has a minimum value of 1000 and a maximum value of 9999, which doesn't make sense and should be investigated further, as well as registration month having a minimum value of 0. Also, it is interesting that the most frequently observed price is $0, meaning that some people are giving their car away for free.

First, I will begin by converting the price column to a numeric data type. Before I do this, I need to remove the dollar sign that is in front of the prices.

In [7]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "").astype(float)
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

After removing the dollar sign in front of the prices, I still got an error when converting that column to a float because there are commas in some of the prices, so I had to remove the commas as well.

I will then repeat this process on the odometer column.

In [8]:
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "").astype(float)
autos["odometer"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64

Since I removed the "km" after the odometer entries, which provided critical information that the odometer readings are in kilometers, I will rename the column to odometer_km.

In [9]:
autos.rename(columns = {"odometer" : "odometer_km"}, inplace = True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fue_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

## Price Analysis

Next, I will examine the price column further.

In [10]:
autos["price"].unique().shape

(2357,)

In [11]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

I can see that the minimum price is \\$0, the maximum price is \$100,000,000, and the average (mean) price is \$9,840. However, even though the maximum price is \$100,000,000, the 75th percentile is only \$7,200, which indicates that there are a few outliers that are driving the mean up. There are over 2,000 unique prices in the data set.

In [12]:
autos["price"].value_counts().head()

0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
Name: price, dtype: int64

The most frequently observed price is \\$0, meaning that 1,421 sellers listed their car for free. The next most common prices are also pretty low dollar amounts, ranging from \$500 to \$2500.

In [13]:
autos["price"].value_counts().sort_index(ascending = True).head()

0.0    1421
1.0     156
2.0       3
3.0       1
5.0       2
Name: price, dtype: int64

I sorted the value counts by the index, which in this case is the price, in order to see the number of observations for the lowest prices. I will also do the same thing in descending order in order to see the number of observations for the highest prices.

In [14]:
autos["price"].value_counts().sort_index(ascending = False).head(10)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64

I can see that the lowest prices have relatively high frequencies, especially \\$0. However, the highest prices have low frequencies, with only one listing at \$99,999,999. There are 3 listings with a price of \$12,345,678, which seems like it might be a joke. If I remove all listings with prices over \$1 million, that would only remove 11 data points, but it would probably adjust the mean significantly and make it closer to the median. 

In [15]:
autos = autos[autos["price"].between(0, 1000000)]
autos["price"].describe()

count     49989.000000
mean       5781.194763
std       11833.263624
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      999999.000000
Name: price, dtype: float64

Removing those 11 data points brought the mean down to \\$5,781, which is still higher than the median of \$2,950, but is significantly more accurate than the original mean of \$9,840.

## Odometer Analysis

Next, I will review the odometer column.

In [16]:
autos["odometer_km"].unique().shape

(13,)

In [17]:
autos["odometer_km"].describe()

count     49989.000000
mean     125737.462242
std       40037.226230
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Based on the descriptive statistics above, I can see that there are only 13 different odometer readings in the data set. This indicates that the people making the ads are rounding the odometer reading to the closest milestone as opposed to reporting exact mileage. The minimum odometer reading is 5000 km, and the maximum is 150,000 km. The average (mean) odometer reading is 125,737 km, and the median odometer reading is 150,000 km.

In [18]:
autos["odometer_km"].value_counts().head()

150000.0    32418
125000.0     5170
100000.0     2168
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64

By far, the most commonly observed odometer reading is 150,000 km with over 32,000 observations having that value, which is consistent with 150,000 km being both the median and the maximum value.

In [19]:
autos["odometer_km"].value_counts().sort_index(ascending = True).head()

5000.0     966
10000.0    264
20000.0    784
30000.0    789
40000.0    818
Name: odometer_km, dtype: int64

Similarly to what I did with price, I will sort the value counts by the index (odometer readings) to see how many observations there are for both the lowest odometer values and highest odometer values.

In [20]:
autos["odometer_km"].value_counts().sort_index(ascending = False).head()

150000.0    32418
125000.0     5170
100000.0     2168
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64

Based on the analysis above, it does not appear that there are any outliers. The lowest observed odometer reading of 5000 km has 966 data points, and the highest observed odometer reading of 150,000 km has over 32,000 data points. Therefore, I will not remove any rows from the data set based on odometer values.

## Date Analysis

Next, I will clean and analyze some of the date columns. I noticed when looking at the descriptive statistics for the data set that the registration month and year are stored as numeric data types, but the date_crawled, ad_created, and last_seen columns are stored as strings in a year-month-day hour:minute:second format. I will therefore extract only the date by taking the first 10 characters of the string for each entry in those columns.

In [21]:
date_crawled = autos["date_crawled"].str[:10]
date_ad_created = autos["ad_created"].str[:10]
date_last_seen = autos["last_seen"].str[:10]

Next, I will count how many observations there are for each unique value in those date columns, but normalize it to a percentage instead of a raw count, and then sort in order of date.

In [22]:
date_crawled.value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.025386
2016-03-06    0.013943
2016-03-07    0.035968
2016-03-08    0.033267
2016-03-09    0.033207
2016-03-10    0.032127
2016-03-11    0.032487
2016-03-12    0.036768
2016-03-13    0.015563
2016-03-14    0.036628
2016-03-15    0.033987
2016-03-16    0.029506
2016-03-17    0.031527
2016-03-18    0.013063
2016-03-19    0.034908
2016-03-20    0.037828
2016-03-21    0.037508
2016-03-22    0.032907
2016-03-23    0.032387
2016-03-24    0.029106
2016-03-25    0.031747
2016-03-26    0.032487
2016-03-27    0.031047
2016-03-28    0.034848
2016-03-29    0.034168
2016-03-30    0.033627
2016-03-31    0.031907
2016-04-01    0.033807
2016-04-02    0.035408
2016-04-03    0.038689
2016-04-04    0.036488
2016-04-05    0.013103
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

I can see that the date the ad was scraped from the German eBay website ranges from March 5th - April 7th, 2016, with a relatively even distribution of how many ads were scraped each day. 

In [23]:
date_ad_created.value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033227
2016-03-10    0.031867
2016-03-11    0.032787
2016-03-12    0.036608
2016-03-13    0.016924
2016-03-14    0.035228
2016-03-15    0.033747
2016-03-16    0.030007
2016-03-17    0.031207
2016-03-18    0.013723
2016-03-19    0.033847
2016-03-20    0.037868
2016-03-21 

The ads were created between June 11, 2015 and April 7th, 2016, with more of the ads seeming to have been made towards the later end of the date range. This makes sense, since items listed for sale probably sell within a few weeks, so it is likely that listings were made within a close time frame of when the scraping was done. It is less likely for a car from almost a year ago to still be listed for sale. 

In [24]:
date_last_seen.value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007582
2016-03-09    0.009842
2016-03-10    0.010762
2016-03-11    0.012523
2016-03-12    0.023805
2016-03-13    0.008982
2016-03-14    0.012803
2016-03-15    0.015883
2016-03-16    0.016444
2016-03-17    0.027926
2016-03-18    0.007422
2016-03-19    0.015743
2016-03-20    0.020705
2016-03-21    0.020725
2016-03-22    0.021585
2016-03-23    0.018584
2016-03-24    0.019564
2016-03-25    0.019204
2016-03-26    0.016964
2016-03-27    0.016024
2016-03-28    0.020865
2016-03-29    0.022345
2016-03-30    0.024845
2016-03-31    0.023825
2016-04-01    0.023105
2016-04-02    0.024885
2016-04-03    0.025366
2016-04-04    0.024625
2016-04-05    0.124267
2016-04-06    0.220989
2016-04-07    0.130949
Name: last_seen, dtype: float64

The date that the ad was last seen ranges from March 5th - April 7th, 2016, the same range as when the ad was scraped, but almost half of the ads were last seen in the last three days of the range: April 5th - 7th. This is likely due to the fact that that is when the scraping stopped and not because more cars were sold on those days.

I will also investigate the registration month and year, since I noticed during my initial data exploration that some values didn't make sense. I will look at the number of observations corresponding to the lowest and highest registration year values, since the minimum of 1000 and the maximum of 9999 don't make sense, and I want to see if there are any other values that don't make sense.

In [25]:
autos["registration_year"].value_counts().sort_index(ascending = False).head(20)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1316
2015     399
2014     665
2013     806
2012    1323
2011    1634
Name: registration_year, dtype: int64

In [26]:
autos["registration_year"].value_counts().sort_index(ascending = True).head(20)

1000    1
1001    1
1111    1
1500    1
1800    2
1910    9
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
Name: registration_year, dtype: int64

There are some registration year values like 1000, 1001, 1111, 1500, and 1800 that don't make sense because cars didn't even exist at that time. The earliest year that is plausible is 1910. On the other end, there are some values like 9999, 9996, 9000, 8888, 6200, 5911, 5000, 4800, 4500, 4100, and 2800 that don't make sense and are not possible as the registration year. In addition, even the registration years of 2017 - 2019 don't make sense, because it is not possible for the car to be registered before the listing was made, and the latest ad was posted on April 7th, 2016.

Therefore, I will remove any data entries with registration years before 1910 or after 2016.

In [27]:
autos = autos[autos["registration_year"].between(1910, 2016)]
autos["registration_year"].describe()

count    48019.000000
mean      2002.805306
std          7.307168
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

I can see that the number of rows in the data set has been reduced from 49,989 to 48,019, and that the mean and the median year for registration is 2003. The minimum year is now 1910 and the maximum year is 2016.

In [28]:
autos["registration_year"].value_counts(normalize = True)

2000    0.069847
2005    0.062788
1999    0.062434
2004    0.056998
2003    0.056790
2006    0.056374
2001    0.056269
2002    0.052750
1998    0.051084
2007    0.047981
2008    0.046461
2009    0.043691
1997    0.042233
2011    0.034028
2010    0.033258
1996    0.030071
2012    0.027552
2016    0.027406
1995    0.027323
2013    0.016785
2014    0.013849
1994    0.013745
1993    0.009267
2015    0.008309
1990    0.008226
1992    0.008122
1991    0.007414
1989    0.003769
1988    0.002957
1985    0.002187
          ...   
1966    0.000458
1977    0.000458
1975    0.000396
1969    0.000396
1965    0.000354
1964    0.000250
1910    0.000187
1963    0.000187
1959    0.000146
1961    0.000125
1956    0.000104
1958    0.000083
1937    0.000083
1962    0.000083
1950    0.000062
1954    0.000042
1941    0.000042
1934    0.000042
1957    0.000042
1951    0.000042
1955    0.000042
1931    0.000021
1953    0.000021
1943    0.000021
1938    0.000021
1939    0.000021
1927    0.000021
1929    0.0000

The most common registration year is 2000, followed by 2005 and then 1999. This makes sense, especially given the relatively low price and high odometer readings that the cars have on average.

In [29]:
autos["registration_month"].value_counts().sort_index(ascending = True)

0     4586
1     3163
2     2914
3     4897
4     3954
5     3949
6     4211
7     3811
8     3085
9     3300
10    3554
11    3271
12    3324
Name: registration_month, dtype: int64

There are over 5,000 data entries in the data set with a month of 0, which doesn't make sense. I would assume that a month of 1 corresponds with January and a month of 12 corresponds with December. It is possible that any listings that only had a registration year but not a month were listed as having a month of 0. Therefore, I will not remove any rows that have 0 specified as the registration month.

## Brand

Finally, I am interested in looking at the most common brands of cars that are listed on German eBay.

In [30]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [31]:
autos["brand"].value_counts(normalize = True)

volkswagen        0.212145
bmw               0.110019
opel              0.108166
mercedes_benz     0.095358
audi              0.086403
ford              0.069785
renault           0.047356
peugeot           0.029530
fiat              0.025865
seat              0.018180
skoda             0.016035
mazda             0.015140
nissan            0.015098
smart             0.013911
citroen           0.013911
toyota            0.012474
sonstige_autos    0.010892
hyundai           0.009850
volvo             0.009246
mini              0.008642
mitsubishi        0.008143
honda             0.007851
kia               0.007101
alfa_romeo        0.006622
porsche           0.006102
suzuki            0.005914
chevrolet         0.005706
chrysler          0.003665
daihatsu          0.002561
dacia             0.002561
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001499
r

For the purposes of this analysis, I will not consider any brands which correspond to less than 1% of the entries in the data set, which means I will consider the top 17 auto brands ranging from Volkswagen to Sonstige Autos.

In [32]:
brands = autos["brand"].value_counts(normalize = True).index[0:17]

I will create a blank dictionary to contain the top 17 brands and their corresponding average listing price. I will then loop over each of the 17 top brands and filter the data set to only the rows corresponding to that brand, then select the price column and find the mean. I will then assign this brand and associated mean price to the dictionary.

In [33]:
price_dict = {}
for entry in brands:
    mean_price = autos.loc[autos["brand"] == entry, "price"].mean()
    price_dict[entry] = mean_price
price_dict

{'audi': 9093.65003615329,
 'bmw': 8102.536248343744,
 'citroen': 3699.935628742515,
 'fiat': 2711.8011272141707,
 'ford': 3949.42345568487,
 'mazda': 4010.7716643741405,
 'mercedes_benz': 8485.239571958942,
 'nissan': 4664.891034482758,
 'opel': 2876.716403542549,
 'peugeot': 3039.4682651622,
 'renault': 2395.4164467897976,
 'seat': 4296.492554410081,
 'skoda': 6334.91948051948,
 'smart': 3542.706586826347,
 'sonstige_autos': 10805.078393881453,
 'toyota': 5115.33388981636,
 'volkswagen': 5426.382546382644}

Based on the results above, I can see that the brand with the highest average listing price is Sonstige Autos at almost 11,000 dollars, followed by Audi, Mercedes Benz, and BMW. The brand with the lowest listing price is Renault.

I am interested to see if the most and least expensive car brands are correlated with the average mileage of those cars. I will repeat the process I used above to come up with the average mileage for the top 17 brands.

In [34]:
mileage_dict = {}
for entry in brands:
    mean_mileage = autos.loc[autos["brand"] == entry, "odometer_km"].mean()
    mileage_dict[entry] = mean_mileage
mileage_dict

{'audi': 129287.78018799711,
 'bmw': 132431.38368351315,
 'citroen': 119580.8383233533,
 'fiat': 116553.94524959743,
 'ford': 124068.93464637421,
 'mazda': 124745.5295735901,
 'mercedes_benz': 130856.0821139987,
 'nissan': 118572.41379310345,
 'opel': 129223.14208702349,
 'peugeot': 127136.81241184767,
 'renault': 128183.81706244503,
 'seat': 121563.57388316152,
 'skoda': 110954.54545454546,
 'smart': 99595.80838323354,
 'sonstige_autos': 87466.53919694072,
 'toyota': 115709.51585976628,
 'volkswagen': 128728.28114263277}

In order to compare these dictionaries, I will convert them both to pandas series and then combine them in a dataframe.

In [35]:
price_series = pd.Series(price_dict)
mileage_series = pd.Series(mileage_dict)
df = pd.DataFrame(price_series, columns = ['mean_price'])
df["mean_mileage"] = mileage_series
df

mean_price   mean_mileage
audi             9093.650036  129287.780188
bmw              8102.536248  132431.383684
citroen          3699.935629  119580.838323
fiat             2711.801127  116553.945250
ford             3949.423456  124068.934646
mazda            4010.771664  124745.529574
mercedes_benz    8485.239572  130856.082114
nissan           4664.891034  118572.413793
opel             2876.716404  129223.142087
peugeot          3039.468265  127136.812412
renault          2395.416447  128183.817062
seat             4296.492554  121563.573883
skoda            6334.919481  110954.545455
smart            3542.706587   99595.808383
sonstige_autos  10805.078394   87466.539197
toyota           5115.333890  115709.515860
volkswagen       5426.382546  128728.281143

From this, I can see that the car brand with the highest average list price, Sonstige Autos, also has the lowest average mileage. However, the other brands with the highest list prices - Audi, Mercedes Benz, and BMW - also have the highest average mileages, so it does not appear that the higher list prices are being driven by lower mileage. The higher prices must be driven by some other factor, like the car brand itself being able to retain value.

## Conclusion

Overall, I was able to eliminate some observations from the data set for which the price or registration year were outliers or didn't make sense. The average listing price of a car on German eBay is roughly \\$5000, most cars have 150,000 km, and the average year that the car was first registered is 2003. The most popular car brands listed for sale are Volkswagen, BMW, and Opel, and certain car brands have higher list prices on average. However, these car brands do not have lower mileage on average.